src/JsonSchema/RefResolver.php

```diff
@@ -63,11 +54,36 @@ public function __construct($retriever = null)
      */
     public function fetchRef($ref, $sourceUri)
     {
-        $retriever  = $this->getUriRetriever();
-        $jsonSchema = $retriever->retrieve($ref, $sourceUri);
-        $this->resolve($jsonSchema);
+        // Get absolute uri
+        $resolver = new UriResolver();
+        $uri = $resolver->resolve($ref, $sourceUri);
+
+        // Split in to location and fragment
+        $location = $resolver->extractLocation($uri);
+        $fragment = $resolver->extractFragment($uri);
+
+        // Retrieve dereferenced schema
+        if ($location == null) {
+            $schema = end($this->schemas);
```

Comment:
I would use a separate variable for storing this. Currently, `$schemas` is used as associative array (map) from Uri => schema object. Using the same data structure also as stack is prone to errors.

We could simply keep `$rootSchema` for this.